### ISSUE TITLE
Memory leak?

### ISSUE THREAD

**Issue by vkuznet** [rana/ora#69]
Hi, I'm trying to understand the memory leak I experience by using ora driver. I see that with every 1k requests I sent concurrent to my app, I have memory increase. The go profiler shows growing number of to ora *Rset, *DrvStmt, etc. I put heap profile over [here](https://gist.github.com/vkuznet/cc438bd03bfe857b8d68). The code I execute follows practices I read on Go web site, i.e. I close Rows, I use single *sql.DB, etc. For reference my execute function is [here](https://github.com/vkuznet/dbs2go/blob/master/src/dbs/dbs.go#L168). But I still struggle to understand why memory leaks. I'm new to Go database/sql and obviously may miss something.
Any advise?
Thanks, Valentin

**Comment by tgulacsi** [comment]
I have never understood the hand rolled free list (why not a sync.Pool),
maybe we have to drill into it.

Valentin Kuznetsov <notifications@github.com> ezt írta (időpont: 2016.
márc. 9., Sze 17:50):

**Comment by rana** [comment]
Sync.pool was originally used then switched to hand-rolled to control
lifetime of the resource. Sync.pool releases resources at its discretion.
Am open to an alternative approach.

**Comment by vkuznet** [comment]
Can you point me where the hand-rolled code resides and how to substitute it with
Sync.pool I can try to tweak the code and see how it goes.

**Comment by rana** [comment]
Hi Valentin,

Here are a couple pool spots to review.

type pool declaration
https://github.com/rana/ora/blob/master/type.go#L736

pool usage
https://github.com/rana/ora/search?utf8=%E2%9C%93&q=pool


Thank you,

Rana

On Wed, Mar 9, 2016 at 2:34 PM, Valentin Kuznetsov <notifications@github.com
> wrote:

> Can you point me where the hand-rolled code resides and how to substitute
> it with
> Sync.pool I can try to tweak the code and see how it goes.
>

**Comment by tgulacsi** [comment]
PTAL the "pool" branch!

I've created a pooler interface, renamed pool to listPool, added syncPool, changed the default "newPool" to be newSyncPool, have left the errList as newListPool.

You can change the default pool implementation at type.go#740

Maximum memory usage of "go test" didn't change significantly, but it does varied during the test - it releases memory back to the OS.

Please test it!

**Comment by vkuznet** [comment]
Hi, I perform some tests with new code and it does not look that it improved memory wise. I checkout new code, put it in place in my GOPATH, run go get again to verify that it can link with Oracle sdk. Then re-compiled my code and run it again. Here is a snapshot of continuous runs where I fetched 1k records from my app and look at top profile after each iteration. I also attach the similar profile for oci8 oracle driver I found on Go drivers page:

```
ora driver

 1357 valya     20   0 2752m 106m  15m S  4.0  1.3   0:04.91 dbs2go
 1357 valya     20   0 2752m 110m  15m S  4.0  1.4   0:06.99 dbs2go
 1357 valya     20   0 2752m 114m  15m S  4.0  1.5   0:09.12 dbs2go
 1357 valya     20   0 2752m 119m  15m S  2.0  1.5   0:11.26 dbs2go
 1357 valya     20   0 2753m 123m  15m S  0.0  1.6   0:13.28 dbs2go
 1357 valya     20   0 2823m 128m  15m S  2.0  1.6   0:15.41 dbs2go
 1357 valya     20   0 2894m 133m  15m S  2.0  1.7   0:17.57 dbs2go
 1357 valya     20   0 2894m 138m  15m S  0.0  1.8   0:19.76 dbs2go
 1357 valya     20   0 2894m 143m  15m S  0.0  1.8   0:21.92 dbs2go
 1357 valya     20   0 2895m 147m  15m S  2.0  1.9   0:24.03 dbs2go
 1357 valya     20   0 2895m 151m  15m S  0.0  1.9   0:26.26 dbs2go
 1357 valya     20   0 2895m 155m  15m S  0.0  2.0   0:28.53 dbs2go
 1357 valya     20   0 2895m 160m  15m S  0.0  2.0   0:30.70 dbs2go
 1357 valya     20   0 2895m 164m  15m S  2.0  2.1   0:32.95 dbs2go
 1357 valya     20   0 2896m 168m  15m S  2.0  2.1   0:35.19 dbs2go
 1357 valya     20   0 2896m 173m  15m S  0.0  2.2   0:37.37 dbs2go
 1357 valya     20   0 2896m 177m  15m S  0.0  2.3   0:39.60 dbs2go
 1357 valya     20   0 2896m 181m  15m S  2.0  2.3   0:41.83 dbs2go
 1357 valya     20   0 2896m 186m  15m S  0.0  2.4   0:44.01 dbs2go
 1357 valya     20   0 2896m 190m  15m S  0.0  2.4   0:46.24 dbs2go
 1357 valya     20   0 2896m 194m  15m S  6.0  2.5   0:48.54 dbs2go
 1357 valya     20   0 2896m 198m  15m S  2.0  2.5   0:50.80 dbs2go
 1357 valya     20   0 2896m 203m  15m S  2.0  2.6   0:52.97 dbs2go
 1357 valya     20   0 2897m 207m  15m S  0.0  2.6   0:55.22 dbs2go
 1357 valya     20   0 2897m 211m  15m S  0.0  2.7   0:57.40 dbs2go
 1357 valya     20   0 2897m 216m  15m S  0.0  2.7   0:59.71 dbs2go
 1357 valya     20   0 2897m 220m  15m S  0.0  2.8   1:01.88 dbs2go
 1357 valya     20   0 2897m 224m  15m S  4.0  2.9   1:03.98 dbs2go
 1357 valya     20   0 2897m 229m  15m S  0.0  2.9   1:06.13 dbs2go
 1357 valya     20   0 2898m 233m  15m S  2.0  3.0   1:08.23 dbs2go

oci8 driver:

 1845 valya     20   0 3327m 110m  15m S  0.0  1.4   0:03.02 dbs2go
 1845 valya     20   0 3330m 112m  15m S  0.0  1.4   0:03.99 dbs2go
 1845 valya     20   0 3330m 112m  15m S  0.0  1.4   0:04.96 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:05.90 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:06.85 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:07.84 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:08.81 dbs2go
 1845 valya     20   0 3330m 113m  15m S  4.0  1.4   0:09.76 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:10.74 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:11.73 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:12.70 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:13.65 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:14.63 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:15.60 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:16.56 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:17.57 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:18.56 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:19.54 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:20.52 dbs2go
 1845 valya     20   0 3330m 113m  15m S  2.0  1.4   0:21.50 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:22.52 dbs2go
 1845 valya     20   0 3330m 113m  15m S  2.0  1.4   0:23.48 dbs2go
 1845 valya     20   0 3330m 113m  15m S  2.0  1.4   0:24.47 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:25.45 dbs2go
 1845 valya     20   0 3330m 113m  15m S  2.0  1.4   0:26.41 dbs2go
 1845 valya     20   0 3330m 113m  15m S  4.0  1.4   0:27.38 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:28.35 dbs2go
 1845 valya     20   0 3330m 113m  15m S  0.0  1.4   0:29.57 dbs2go
 1845 valya     20   0 3330m 113m  15m S  2.0  1.4   0:30.81 dbs2go
 1845 valya     20   0 3330m 113m  15m S  2.0  1.4   0:32.20 dbs2go
```

Please let me know if I need to tune something. Thanks. Valentin.

**Comment by tgulacsi** [comment]
Strange.

Can you put up the heap profile _svg_ somewhere? Without the exact binary, the textual format does not say anything.
What kind of columns does the query return? How many rows?
Can you create a test case?

**Comment by vkuznet** [comment]
I'm attaching now two svg's. One I took when client made requests, and another when server an idle for a while. SVG when clients talk to server: https://www.dropbox.com/s/fzaczs889szb4h2/client_server.svg?dl=0 and here is idle server: https://www.dropbox.com/s/fqtfk1v5xfi46h3/idle_server.svg?dl=0

Also here is current heap (i.e. after clients made all the load and server stayed for a while): https://gist.github.com/vkuznet/4b6e04b97e152d08002d

After running 20 times with 1k request per round I end-up with 255MB of RAM. As I posted before each round clients sent 1k requests and RAM increases by around 5MB.

The query I used in tests can be found here: https://github.com/vkuznet/dbs2go/blob/master/static/sql/datasets.sql
there is about 3000 rows I fetch each round. The columns are mixed types, int and strings. I use live data, but I don't have rights on our Oracle db for making my own table to do the test.

Let me know if you want anything else.

**Comment by tgulacsi** [comment]
Sorry, I cannot reproduce.

I've inserted a "BenchmarkMemory" test function into z_oracle_test.go, in "issue69" branch.
Please try to tweak the query, to mimic your, or at least to reproduce the problem!
For me ora.v3.test uses a constant 30MB, no fluctuations.


I couldn't understand the given profiles, either: client_server.svg says 8MB is in use, of which 2+2.5MB is by the HTTP server, 2MB is dbs.datasets
idle_server.svg says 0B is in use.

The go tool pprf needs the running binary, not just the profile.

What is your go's version? I'm using go1.6.

**Comment by vkuznet** [comment]
Tamás, I adopted your z_oracle_test.go to structure I used in my code, i.e. I put the same function, structures, query I used in my code such that you can inspect everything.  For simplicity I removed in my local setup all other tests. Then I run your test as following:
```
go test -c && ./ora.v3.test -test.run=^$ -test.bench=Memory -test.memprofilerate=1 -test.memprofile=/tmp/mem.prof && go tool pprof --alloc_space ora.v3.test /tmp/mem.prof
```

The adopted z_oracle_test.go code, output, top10 and svg files can be found here: https://www.dropbox.com/sh/9k8zbd7yhx2zz36/AAB_K1UBVTzqd5WMHQTO85kEa?dl=0

Please let me know if you need anything from my side.

**Comment by tgulacsi** [comment]
The query column's types and sizes would be great, although from the profile, I assume that some numbers are returned as strings, and that may allocate a lot.

**Comment by tgulacsi** [comment]
The main problem is with NUMBER columns (non-restricted, e.g. 38 digits at max).
As these can hold 38 digits, and that can't be represented as int64, we use the ora.N column type, which returns these as strings. As long as I see, this is the main source of allocations.

PTAL "issue69" branch's ce36609 z_benchmem_test.go - with this commit in place you can tweak the used column type. If you _know_ that your columns can't be bigger than what can be represented as an int64, use SetNumberBigInt(ora.I64); for float64, you can use SetNumberBigFloat(ora.F64).

Hope this helps.

**Comment by tgulacsi** [comment]
Another possible solution: use what Oracle provides: OCINumber.

I've implemented its encoding/decoding to/from string and wired it in to use for result sets.
This means another change (not .(string) but .(OCINum).String()), but stores only the significant digits, at most 21 bytes!

Please try the "ocinum" branch!

**Comment by vkuznet** [comment]
Tamás,
I'll use both of your suggestion, but can look at them only next week.
I appreciate your effort and will try to help as much as I can.
Thanks,
Valentin.

**Comment by vkuznet** [comment]
Tamás, I got some time to test stuff. I tried ocinum branch but I can't compile anything since "github.com/rana/ora/num" does not exist (probably you didn't upload it). This is required in here: https://github.com/rana/ora/blob/ocinum/defOCINum.go#L15

**Comment by tgulacsi** [comment]
Please update to 5a1c460f59c59d43e9b12fd9c370c1963ac62933 I've pushed the fix for the imports.
Sorry for the inconvenience!

**Comment by vkuznet** [comment]
Tamás, I further debug the problem and now I think I understand what's going on. Yes, the problem relates to strings used for casting values. I used explicit columns types in Scan and problem disappeared once I switched to sql.NullInt64, sql.NullFloat64 and (most importantly) to sql.NullString. Apart from NUMBERs you mentioned I found that using string data types still provide memory leak. Only when I switched to sql.NullString data type I no longer see the leak I saw before. Also, the memory footprint significantly reduced by factor of 2 and all my tests runs at 52MB. And, for the reference I used issue69 branch you provided. You may close the ticket now. Thanks for your help.

**Comment by tgulacsi** [comment]
I've merged issue69 to allow bigger flexibility. Still would love to see some tests with the "ocinum" branch - even a "working, no bigger memory usage" would suffice!

**Comment by vkuznet** [comment]
Give me some time, I'll try to do that.

**Comment by vkuznet** [comment]
Tamás, I tried my code with ora-ocinum branch and it's working fine now, no more memory leaks. Feel free to close the ticket. Thanks. Valentin.

**Comment by tgulacsi** [comment]
Thanks! Merged it to master, will be the next v3.4.
